compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -291,167 +358,209 @@ class Semantic {
               val error = AccessNonInit(target, trace.add(source).toVector)
               Result(Hot, error :: Nil)
 
-        case _: Fun =>
-          ???
+        case fun: Fun =>
+          report.error("unexpected tree in selecting a function, fun = " + fun.expr.show, source)
+          Result(Hot, Nil)
 
         case RefSet(refs) =>
           val resList = refs.map(_.select(field, source))
           val value2 = resList.map(_.value).join
           val errors = resList.flatMap(_.errors)
           Result(value2, errors)
       }
+    }
 
-    def call(meth: Symbol, superType: Type, source: Tree, needResolve: Boolean = true): Contextual[Result] =
-      value match {
+    def call(meth: Symbol, args: List[ArgInfo], superType: Type, source: Tree, needResolve: Boolean = true): Contextual[Result] = log("call " + meth.show + ", args = " + args, printer, res => res.asInstanceOf[Result].show) {
+      def checkArgs = args.flatMap(_.promote)
+
+      // fast track if the current object is already initialized
+      if promoted.isCurrentObjectPromoted then Result(Hot, Nil)
+      else value match {
         case Hot  =>
-          Result(Hot, Errors.empty)
+          Result(Hot, checkArgs)
 
         case Cold =>
           val error = CallCold(meth, source, trace.toVector)
-          Result(Hot, error :: Nil)
+          Result(Hot, error :: checkArgs)
 
         case addr: Addr =>
+          val isLocal = meth.owner.isClass
           val target =
             if !needResolve then
               meth
             else if superType.exists then
               resolveSuper(addr.klass, superType, meth)
             else
               resolve(addr.klass, meth)
-          if target.isOneOf(Flags.Method | Flags.Lazy) then
+
+          if target.isOneOf(Flags.Method) then
+            val trace1 = trace.add(source)
             if target.hasSource then
+              given Trace = trace1
               val cls = target.owner.enclosingClass.asClass
+              val ddef = target.defTree.asInstanceOf[DefDef]
+              val env2 = Env(ddef, args.map(_.value).widen)
               if target.isPrimaryConstructor then
+                given Env = env2
```

Comment:
Is there a specific reason to prefer this pattern of `given` over the `use()` method?

Reply:
I try to use `given T = e` whenever possible. However, in same places we cannot easily use `given`, because the scoping interference. I described the problem in more detail [here](https://github.com/lampepfl/dotty/discussions/12529).